File: framework/languages/dotnet/samples/SupportChat/Server/Support/Adapters/ZLink/Spots/ConversationSpot.cs
using Systems.Zlink;
using Systems.Zlink.Codecs.Json;
using SupportChat.Server.Support.Adapters.ZLink.Actors;
using SupportChat.Server.Support.Adapters.ZLink.Notifications;
using SupportChat.Server.Support.Adapters.ZLink.Spots.Handlers;
using SupportChat.Server.Support.Domain.SupportChat;
using SupportChat.Server.Configuration;
using SupportChat.Shared.Contracts;
using Microsoft.Extensions.Logging;
using Zlink.Framework.Contracts.Spots;

namespace SupportChat.Server.Support.Adapters.ZLink.Spots;

internal sealed class ConversationSpot(
    IZLinkSpotContext context,
    ConversationNotificationPublisher notifications,
    ILogger<ConversationSpot> logger) : IZLinkSpot<SupportUserActor>
{
    internal static readonly TimeSpan IdleCheckPeriod = TimeSpan.FromMilliseconds(200);

    private readonly Dictionary<string, SupportUserActor> _actors = new(StringComparer.Ordinal);
    private Conversation? _conversation;

    public IZLinkSpotContext Context { get; } = context;

    public void Configure()
    {
        Context.Handlers.AddHandler<SendChatMessageHandler>();
        Context.Handlers.AddHandler<SetTypingHandler>();
        Context.Handlers.AddHandler<CloseConversationHandler>();
    }

    public async ValueTask<ZLinkSpotCreateResponse> OnCreateAsync(
        Message request,
        CancellationToken cancellationToken)
    {
        var create = request.FromJson<ConversationCreateRequest>();
        var conversationId = Context.SpotRid.ToHex();
        _conversation = new Conversation(
            conversationId,
            create.Subject,
            create.CustomerActorId,
            create.CustomerDisplayName,
            create.CreatedAtUnixMs,
            new ConversationPolicy(
                SampleTimings.IdleTimeout,
                SampleTimings.CloseGraceTimeout,
                MaxMessageLength: 500));
        _ = cancellationToken;
        logger.LogInformation(
            "support conversation: created. conversation={ConversationId}, customer={CustomerActorId}",
            conversationId,
            create.CustomerActorId);
        Console.Error.WriteLine($"support conversation: created conversation={conversationId} customer={create.CustomerActorId}");
        return ZLinkSpotCreateResponse.Accept();
    }

    public async ValueTask OnClosingAsync(CancellationToken cancellationToken)
    {
        _ = cancellationToken;
        await ValueTask.CompletedTask;
    }

    public async ValueTask<ZLinkSpotActorJoinResult> OnActorJoinAsync(
        SupportUserActor actor,
        Message request,
        CancellationToken cancellationToken)
    {
        var join = request.Decode<JoinConversationReq>();
        var conversation = RequireConversation();
        if (!string.Equals(join.ConversationId, conversation.ConversationId, StringComparison.Ordinal))
        {
            return ZLinkSpotActorJoinResult.Reject();
        }

        if (string.Equals(actor.Role, SupportChatRoles.Agent, StringComparison.Ordinal))
        {
            var agentState = await JoinAgentAsync(actor, cancellationToken);
            return ZLinkSpotActorJoinResult.Accept(new JoinConversationRes(agentState).Encode());
        }

        actor.JoinConversation(join.ConversationId);
        _actors[actor.ActorId] = actor;
        if (string.Equals(actor.ActorId, conversation.CustomerActorId, StringComparison.Ordinal))
        {
            await notifications.PublishJoinedAgentToCustomerAsync(
                actor,
                conversation.Snapshot(),
                cancellationToken);
        }

        logger.LogInformation(
            "support conversation: actor joined. conversation={ConversationId}, actor={ActorId}, role={Role}",
            join.ConversationId,
            actor.ActorId,
            actor.Role);
        Console.Error.WriteLine($"support conversation: actor joined conversation={join.ConversationId} actor={actor.ActorId} role={actor.Role}");
        return ZLinkSpotActorJoinResult.Accept(new JoinConversationRes(conversation.Snapshot()).Encode());
    }

    public async ValueTask<ConversationState> JoinAgentAsync(
        SupportUserActor agent,
        CancellationToken cancellationToken)
    {
        var conversation = RequireConversation();
        var change = conversation.JoinAgent(agent.ActorId, agent.DisplayName, NowUnixMs());
        agent.JoinConversation(conversation.ConversationId);
        _actors[agent.ActorId] = agent;
        await notifications.PublishAsync(change.Events, _actors, cancellationToken);
        return change.State;
    }

    public async ValueTask<SendChatMessageRes> SendMessageAsync(
        SupportUserActor actor,
        SendChatMessageReq request,
        CancellationToken cancellationToken)
    {
        EnsureConversationId(request.ConversationId);
        var change = RequireConversation().SendMessage(actor.ActorId, request.Text, NowUnixMs());
        await notifications.PublishAsync(change.Events, _actors, cancellationToken);
        ScheduleIdleCheck();
        return new SendChatMessageRes(
            change.Events.Single(static item => item.Kind == ConversationEventKind.MessageAppended).Message
                ?? throw new InvalidOperationException("Message event was not created."),
            change.State);
    }

    public async ValueTask<SetTypingRes> SetTypingAsync(
        SupportUserActor actor,
        SetTypingReq request,
        CancellationToken cancellationToken)
    {
        EnsureConversationId(request.ConversationId);
        var change = RequireConversation().SetTyping(actor.ActorId, request.IsTyping);
        await notifications.PublishAsync(change.Events, _actors, cancellationToken);
        return new SetTypingRes(change.State);
    }

    public async ValueTask<CloseConversationRes> CloseAsync(
        SupportUserActor actor,
        CloseConversationReq request,
        CancellationToken cancellationToken)
    {
        EnsureConversationId(request.ConversationId);
        var change = RequireConversation().Close(actor.ActorId, request.Reason);
        await notifications.PublishAsync(change.Events, _actors, cancellationToken);
        return new CloseConversationRes(change.State);
    }

    internal async ValueTask CheckIdleAsync(CancellationToken cancellationToken)
    {
        var change = RequireConversation().MarkIdle(NowUnixMs());
        await notifications.PublishAsync(change.Events, _actors, cancellationToken);
    }

    private void EnsureConversationId(string conversationId)
    {
        var conversation = RequireConversation();
        if (!string.Equals(conversationId, conversation.ConversationId, StringComparison.Ordinal))
        {
            throw new InvalidOperationException($"Request targets another conversation. conversation={conversationId}");
        }
    }

    private Conversation RequireConversation()
    {
        return _conversation ?? throw new InvalidOperationException("Conversation has not been created.");
    }

    private static long NowUnixMs()
    {
        return DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
    }

    private void ScheduleIdleCheck()
    {
        _ = Task.Run(async () =>
        {
            await Task.Delay(SampleTimings.IdleTimeout);
            while (true)
            {
                var conversation = RequireConversation();
                var change = conversation.MarkIdle(NowUnixMs());
                await notifications.PublishAsync(change.Events, _actors, CancellationToken.None);
                if (change.Events.Count > 0
                    || !string.Equals(change.State.Status, ConversationStatuses.Active, StringComparison.Ordinal)
                    || change.State.IdleDeadlineUnixMs is null)
                {
                    return;
                }

                await Task.Delay(IdleCheckPeriod);
            }
        });
    }
}

File: framework/languages/node/samples/SupportChat.Ts/Server/Support/Infrastructure/ZLink/Spots/conversation-spot.ts
import { Injectable } from '@nestjs/common';
import { Conversation } from '../../../Domain/SupportChat/conversation';
import { SupportNotificationPublisher } from '../Notifications/support-notification-publisher';
import { SupportChatRoles, joinConversationRes } from '../../../../../Shared/Contracts/messages';
import type {
  ZLinkSpot,
  ZLinkSpotActorJoinResponse,
  ZLinkSpotContext
} from '@zlink-systems/framework';
import type { SupportUserActor as SupportUserActorType } from '../Actors/support-user-actor';
import type { ConversationCreateRequest } from './conversation-create-request';
import type {
  CloseConversationReq,
  CloseConversationRes,
  ConversationState,
  SendChatMessageReq,
  SendChatMessageRes,
  SetTypingReq,
  SetTypingRes
} from '../../../../../Shared/Contracts/messages';

type SupportNotificationPublisherLike = Pick<
  SupportNotificationPublisher,
  'publish' | 'publishJoinedAgentToCustomer'
>;

// ConversationSpot owns the ZLink Spot lifecycle, actor join callbacks, the idle timer and the
// connection to the notification publisher. Domain judgement (message sequence, typing, idle and
// close transitions) lives in Conversation; this adapter only translates callbacks and publishes
// the events Conversation returns.
@Injectable()
class ConversationSpot implements ZLinkSpot<SupportUserActorType> {
  private static notifications: SupportNotificationPublisherLike | null = null;
  readonly context!: ZLinkSpotContext<SupportUserActorType, ConversationSpot>;
  private conversationId = 'supportchat-conversation';
  private conversation: Conversation | null = null;
  private readonly actors = new Map<string, SupportUserActorType>();

  static useNotifications(notifications: SupportNotificationPublisherLike): void {
    ConversationSpot.notifications = notifications;
  }

  async onInitialize(): Promise<void> {
    this.conversationId = this.context.spotRid;
  }

  // The conversation create request is delivered by SupportConversationAllocator immediately
  // after the Spot is created, before any actor joins.
  initializeConversation(create: ConversationCreateRequest): void {
    this.conversationId = this.context.spotRid;
    this.conversation = new Conversation(
      this.conversationId,
      create.subject,
      create.customerActorId,
      create.customerDisplayName,
      create.createdAtUnixMs
    );
  }

  async onActorJoin(actor: SupportUserActorType): Promise<ZLinkSpotActorJoinResponse> {
    try {
      const conversation = this.requireConversation();
      if (actor.role === SupportChatRoles.agent) {
        const state = await this.joinAgent(actor);
        return { accepted: true, reply: joinConversationRes(state) };
      }

      actor.joinConversation(conversation.conversationId);
      this.actors.set(actor.actorId, actor);
      const snapshot = conversation.snapshot();
      // The customer joins after the agent was already assigned, so it never observed the
      // agent's ParticipantJoinedNotify. Deliver a catch-up notify so the customer sees the
      // already-present agent and the Active status.
      if (snapshot.agentActorId !== null) {
        await this.requireNotifications().publishJoinedAgentToCustomer(actor.actorId, snapshot);
      }
      return { accepted: true, reply: joinConversationRes(snapshot) };
    } catch (error) {
      return {
        accepted: false,
        reply: { error: error instanceof Error ? error.message : String(error) }
      };
    }
  }

  async onJoinedActor(actor: SupportUserActorType): Promise<void> {
    void actor;
  }

  async onLeaveActor(actor: SupportUserActorType): Promise<void> {
    this.actors.delete(actor.actorId);
  }

  async onDisconnectActor(actor: SupportUserActorType): Promise<void> {
    actor.markDisconnected();
  }

  private async joinAgent(agent: SupportUserActorType): Promise<ConversationState> {
    const conversation = this.requireConversation();
    const change = conversation.joinAgent(agent.actorId, agent.displayName, Date.now());
    agent.joinConversation(conversation.conversationId);
    this.actors.set(agent.actorId, agent);
    await this.requireNotifications().publish(change.events, this.participantActorIds());
    return change.state;
  }

  async sendMessage(actor: SupportUserActorType, request: SendChatMessageReq): Promise<SendChatMessageRes> {
    this.ensureConversationId(request.conversationId);
    const change = this.requireConversation().sendMessage(actor.actorId, request.text, Date.now());
    await this.requireNotifications().publish(change.events, this.participantActorIds());
    const appended = change.events.find((event) => event.message !== undefined);
    if (appended?.message === undefined) {
      throw new Error('Message event was not created.');
    }
    return { message: appended.message, state: change.state };
  }

  async setTyping(actor: SupportUserActorType, request: SetTypingReq): Promise<SetTypingRes> {
    this.ensureConversationId(request.conversationId);
    const change = this.requireConversation().setTyping(actor.actorId, request.isTyping);
    await this.requireNotifications().publish(change.events, this.participantActorIds());
    return { state: change.state };
  }

  async close(actor: SupportUserActorType, request: CloseConversationReq): Promise<CloseConversationRes> {
    this.ensureConversationId(request.conversationId);
    const change = this.requireConversation().close(actor.actorId, request.reason);
    await this.requireNotifications().publish(change.events, this.participantActorIds());
    return { state: change.state };
  }

  // The idle timer handler only forwards the time signal. Conversation decides whether the
  // conversation becomes WaitingForClose or Closed.
  async checkIdle(): Promise<void> {
    const conversation = this.conversation;
    if (conversation === null) {
      return;
    }
    const now = Date.now();
    const idleChange = conversation.markIdle(now);
    if (idleChange.events.length > 0) {
      await this.requireNotifications().publish(idleChange.events, this.participantActorIds());
    }
  }

  snapshot(): ConversationState {
    return this.requireConversation().snapshot();
  }

  private participantActorIds(): string[] {
    return [...this.actors.keys()];
  }

  private ensureConversationId(conversationId: string): void {
    const conversation = this.requireConversation();
    if (conversationId !== conversation.conversationId) {
      throw new Error(`Request targets another conversation. conversation=${conversationId}`);
    }
  }

  private requireConversation(): Conversation {
    if (this.conversation === null) {
      throw new Error('Conversation has not been created.');
    }
    return this.conversation;
  }

  private requireNotifications(): SupportNotificationPublisherLike {
    if (ConversationSpot.notifications === null) {
      throw new Error('ConversationSpot notifications were not configured.');
    }
    return ConversationSpot.notifications;
  }
}

export { ConversationSpot };
export type { ConversationState };
